src/test/kotlin/com/terraformation/backend/db/DatabaseBackedTest.kt

```diff
@@ -0,0 +1,2482 @@
+import org.springframework.test.context.ActiveProfiles
+import org.springframework.test.context.ContextConfiguration
+import org.springframework.test.context.support.TestPropertySourceUtils
+import org.springframework.transaction.annotation.Transactional
+import org.testcontainers.containers.Network
+import org.testcontainers.containers.PostgreSQLContainer
+import org.testcontainers.junit.jupiter.Testcontainers
+import org.testcontainers.utility.DockerImageName
+
+/**
+ * Superclass for tests that make use of a database. You will usually want to subclass
+ * [DatabaseTest] or [ControllerIntegrationTest] instead of this.
+ *
+ * Base class for database-backed tests. Subclass this to get a fully-configured database with a
+ * [DSLContext] and a set of jOOQ DAO objects ready to use. The database is run in a Docker
+ * container which is torn down after all tests have finished. This cuts down on the chance that
+ * tests will behave differently from one development environment to the next.
+ *
+ * In general, you should only use this for testing database-centric code! Do not put SQL queries in
+ * the middle of business logic. If you want to test code that _uses_ data from the database, pull
+ * the queries out into data-access classes (the code base uses the suffix "Store" for these
+ * classes) and stub out the stores. Then test the store classes separately. Database-backed tests
+ * are slower and are usually not as easy to read or maintain.
+ *
+ * Some things to be aware of:
+ * - Each test method is run in a transaction which is rolled back afterwards, so no need to worry
+ *   about test methods polluting the database for each other if they're writing values.
+ * - But that means test methods can't use data written by previous methods. If your test method
+ *   needs sample data, either put it in a migration (migrations are run before any tests) or in a
+ *   helper method.
+ * - Sequences, including the ones used to generate auto-increment primary keys, are normally not
+ *   reset when transactions are rolled back. But it is useful to have a predictable set of IDs to
+ *   compare against. So subclasses can override [tablesToResetSequences] with a list of tables
+ *   whose primary key sequences should be reset before each test method.
+ */
+@ActiveProfiles("test")
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
+@ContextConfiguration(
+    initializers = [DatabaseBackedTest.DockerPostgresDataSourceInitializer::class])
+@EnableConfigurationProperties(TerrawareServerConfig::class)
+@Testcontainers
+@Transactional
+@ComponentScan(basePackageClasses = [UsersDao::class])
+abstract class DatabaseBackedTest {
+  @Autowired lateinit var dslContext: DSLContext
+
+  /**
+   * List of tables from which sequences are to be reset before each test method. Sequences used
+   * here belong to the primary key in the table.
+   */
+  protected open val tablesToResetSequences: List<Table<out Record>>
```

Comment:
I had to open this value to override it in child classes. I am not really sure why it is implicitly open in `pdd-backend`

Reply:
I thought protected implied open!

Reply:
Yea I actually ended up being able to remove this later when I added the `Transactional` annotation so perhaps it isn't needed here after all. Going to try to remove it in this PR.

Reply:
![SCR-20240603-nopm.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/wJnPX9wLZAkiGXEBFJxL/8da35e48-6bc8-447a-b0ab-8b95ae0ef105.png)

Reply:
Can it be abstract val perhaps? I also think open is good since it gives clear indication that it is meant to be overriden. abstract further confirms that.

Reply:
Problem with abstract is that it is initialized with a default value

Reply:
Makes sense. I think I just have gaps in my Kotlin understanding! `open` seems totally fine and thanks for explaining it!

Reply:
To be honest I don't really understand why it "just works" with the `@Transactional` annotation so that is the thing that is puzzling me

Reply:
It "just works" because we use a compiler plugin that automatically opens variables/functions in classes with certain Spring annotations. That's needed (or at least helpful) because Kotlin makes things `final` by default, but Spring needs to be able to construct subclasses on the fly in order to implement stuff like `@Transactional`. The way `@Transactional` works is that Spring basically builds a little proxy class that wraps each of the methods with code to manage the transactions.

Making this `open` is the right fix here.

Reply:
Ah, I see, thanks for the explainer